### PR TITLE
926 distributed training tests

### LINKED ISSUE #926: Add support to run distributed training tests in CI
**Is your feature request related to a problem? Please describe.**
Now we already have some distributed training test cases, like: https://github.com/Project-MONAI/MONAI/blob/master/tests/test_handler_rocauc_dist.py
Need to be executed in our CI system.
We can run with **2 GPUs in 1 node** first.

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Fixes #926

### Description
adds a utility for distributed tests

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by wyli** [comment]
- tested single machine multiprocess on macos and ubuntu w/o GPU
- implemented some utilities mainly for pickling test cases for multiprocessing.  the decorator pattern could also be useful for https://github.com/Project-MONAI/MONAI/issues/852

on windows CI there's an error for the distributed test `OSError: [WinError 1455] The paging file is too small for this operation to complete. Error loading "C:\hostedtoolcache\windows\Python\3.8.6\x64\lib\site-packages\torch\lib\caffe2_detectron_ops_gpu.dll" or one of its dependencies.` I think this is a limitation of the CI instance rather than a code issue. so I skip the tests on windows

File: tests/utils.py [hunk 1]
```diff
@@ -87,6 +92,169 @@ def make_nifti_image(array, affine=None):
...
+    def setUp(self) -> None:
+        self.original_mp = torch.multiprocessing.get_start_method(allow_none=True)
+        try:
+            torch.multiprocessing.set_start_method("spawn", force=True)
+        except RuntimeError:
+            pass
+
+    def tearDown(self) -> None:
+        try:
+            torch.multiprocessing.set_start_method(str(self.original_mp), force=True)
```
Comment (Nic-Ma) [inline tests/utils.py:109]: Hi @ericspod ,

This is not the recommended `launch method` in PyTorch distributed data parallel, do you think it's OK to test with it? I am not very sure about it. Others looks good to me.

Thanks.

Comment (wyli) [reply tests/utils.py:109]: Hi @Nic-Ma, the main difference is that `torch.distributed.launch` is a simple wrapper of
```py
process = subprocess.Popen(cmd, env=current_env, stdout=stdout_handle, stderr=stderr_handle)
```
this PR's test util is doing
```py
p = torch.multiprocessing.Process(target=self.run_process, args=(func, proc_rank, args, kwargs, results))
```
I agree that `torch.distributed.launch` is closer to the real-world use cases,
but the test util is easier to use for wrapping the test cases, testing out the basic gloo/nccl backend usages

**Review by Nic-Ma — APPROVED** [review]
Let's use `spawn` for now, good enough for the first version, maybe some expert can help update it later.

Thanks.

**Comment by wyli** [comment]
> Let's use `spawn` for now, good enough for the first version, maybe some expert can help update it later.
> 
> Thanks.

sure thanks! this also needs to be extended to multi-node test cases

**Pull request merged**